tools/donate-cpu-server.py

```diff
@@ -255,8 +255,11 @@ def diffMessageIdTodayReport(resultPath, messageId):
 
 def timeReport(resultPath):
     text = 'Time report\n\n'
+    text += 'Great difference between versions:\n'
     text += 'Package ' + OLD_VERSION + ' Head\n'
 
+    textLongRunning = ''
+    longRunningThreshold = 60.0 * 30.0
```

Comment:
hmm.. somehow it would be better if this was depending on how complex/large the project is

Reply:
Yes absolutely. But i have no idea how to implement this in a simple way that does not depend on other tools but is still reliable (for example counting the lines of code without comments?).

We could simply store the size in bytes of all the sources that get extracted. The threshold could be one hour per MB (but at least 30 Minutes) or so. We will still get "false positives" i guess.

IMHO 10 hours for not even 2 MB of code is a bit long (observed that during an analysis that ran all night and did not even finish). It is at least worth a look i think.

If anyone has suggestions for the calculation/implementation i would appreciate it.

Reply:
here in the server we can't do much.

in the `donate-cpu.py` we can do a lot more. The only problem is that we need to put some more details in the results report. This is OK.

In general, I think that some `one hour per MB` threshold sounds pretty ok. Maybe we want to increase that in some special cases (when there are large files, templates, ...).

Reply:
An idea; we could run cppcheck one file at a time. We can then compare the analysis time with the size of the `-E` output. If 10KB output takes more than X minute(s) that may be strange..

Reply:
I agree that the client has to be enhanced to write more necessary details in the output results. But i would prefer to have the calculation and threshold in the server so it can be quickly adapted without the need to rerun the analysis for all packages. But i have no strong opinion on that.

The `--showtime=` output could really be interesting. It would be nice if we could add it to the result output, but it is already (too) large. Not sure how the results could be used to find performance issues but i think that could be implemented when the information of `--showtime=` exists.

> An idea; we could run cppcheck one file at a time. We can then compare the analysis time with the size of the -E output. If 10KB output takes more than X minute(s) that may be strange..
This would drop the whole program analysis results. Or do you think of running this additionally to the current analysis? Either way it sounds a bit too resource intensive to me.

Reply:
> This would drop the whole program analysis results. Or do you think of running this additionally to the current analysis? Either way it sounds a bit too resource intensive to me.

I thought we could use `--cppcheck-build-dir`.. but well I realize now that will not work.

Reply:
> I agree that the client has to be enhanced to write more necessary details in the output results. But i would prefer to have the calculation and threshold in the server so it can be quickly adapted without the need to rerun the analysis for all packages. But i have no strong opinion on that.

It depends on how detailed metrics we want. The full `--showtime` output for each file should not be written in the results file.

Total code size / time can be written.

The client can ask for some threshold(s).

> Or do you think of running this additionally to the current analysis?

That is an option. It would not hurt if this is a server option so we can easily turn this on/off on the server.

Reply:
How about .. compile cppcheck with `-DDACA_AT_HOME` and then it will output simple metrics for each file. Size of `-E` output and analysis time. Maybe a few `--showtime` metrics also.

I do not think that we want to have performance metrics for each file in the results but we could have the details for the worst file(s). Maybe max time for file that is less than 10KB. Max time for file that is less than 100KB. Max time for file that is less than 1MB.

Reply:
> How about .. compile cppcheck with `-DDACA_AT_HOME` and then it will output simple metrics for each file. Size of `-E` output and analysis time. Maybe a few `--showtime` metrics also.

IMHO this would be a good approach. So Cppcheck should just print the information to `stdout` and the client script collects the information and appends it to the result output?

Reply:
> IMHO this would be a good approach. So Cppcheck should just print the information to stdout and the client script collects the information and appends it to the result output?

No I do not want to have this info for each file in the results. That would mean millions of redundant timing results. If file1 takes 1 second to analyse then why upload that info to the server.

Reply:
We have plenty of diskspace right now so that is not a problem. I do not want that the results are cluttered.

If all the performance metrics are uploaded they can be compared on the server side. There is a point to have that possibility. You could upload all metrics and save them in a separate file.

Maybe it would be better to save the diff and warnings separately also. We could for instance reduce the response time for the diff report if the diffs are saved in separate files. But well.. this is a separate problem it does not have to be fixed in this PR.

Reply:
There are many great ideas, but i am not sure if i got them all correctly. So i suggest creating tickets with somewhat detailed requirements to avoid implementing things the "wrong" way.
I will create another PR to fix the package names in the time output so they do not contain all the path and close this PR.
Do you have any other (simple) TODOs for the donate cpu scripts which are more straightforward to implement?

Reply:
> Do you have any other (simple) TODOs for the donate cpu scripts which are more straightforward to implement?

hmm. not off the top of my head. I will think! :-)